### PR TITLE
Added AssaultDash

### PR DESCRIPTION
**Pull request by verysecrethero**
Assault Dash 
- shift("SpecialAbility")button + movement button or doubletap movement button to dash!
- dash button
- You can now disable DoubleTapToDash in Input.Ini: DoubleTapToDash=false
  current version:
- you cant jump and and then dash, you have to wait till you get down on the ground
- you cant dash and then jump, you have to wait till the dash is over (0.25sec)
  explanation: jumping enables on air friction, which keeps the velocity in x,z direction constant while you are in the air. hence if you dash and "go into air" you will keep a very high velocity in x,z till you are down on the ground. hence i had to disable the combo jump&dash.
- not yet connected to an animation
- only players with dash component can dash
- dash component currently doesnt hold any values
- cooldowntimer is set to 2.0 sec
- doubletaptimer is set to 0.25sec

### CODE REVIEW

File: include/Game/Systems/PlayerMovementSystem.h [hunk 1]
```diff
@@ -20,4 +20,20 @@ class PlayerMovementSystem : public ImpureSystem, PureSystem
 
     EventRelay<PlayerMovementSystem, Events::PlayerSpawned> m_EPlayerSpawned;
     bool OnPlayerSpawned(Events::PlayerSpawned& e);
+
+    double m_AssaultDashDoubleTapDeltaTime = 0.0f;
+    double m_AssaultDashCoolDownTimer = 0.0f;
+    double m_AssaultDashCoolDownMaxTimer = 3.0f;
+    ImGuiKey m_AssaultDashDoubleTapLastKey = ImGuiKey_Escape;
```
Comment (sippeangelo) [inline include/Game/Systems/PlayerMovementSystem.h]: Why ImGui?

File: include/Game/Systems/PlayerMovementSystem.h [hunk 2]
```diff
@@ -20,4 +20,20 @@ class PlayerMovementSystem : public ImpureSystem, PureSystem
...
+    const float m_AssaultDashDoubleTapSensitivityTimer = 0.25f;
+    enum class AssaultDashDirection {
+        Left,
+        Right,
+        None
+    };
+    AssaultDashDirection m_AssaultDashTapDirection = AssaultDashDirection::None;
+    bool m_AssaultDashDoubleTapped = false;
+    bool m_PlayerIsDashing = false;
```
Comment (sippeangelo) [inline include/Game/Systems/PlayerMovementSystem.h]: This won't work for multiple players

File: src/Game/Systems/PlayerMovementSystem.cpp [hunk 3]
```diff
@@ -71,12 +73,15 @@ void PlayerMovementSystem::Update(double dt)
...
+                //if doubleTapped do Assault Dash - but only boost maximum 50.0f
+                float doubleTapDashBoost = m_AssaultDashDoubleTapped ? 20.0f : 1.0f;
+                accelerationSpeed = glm::min(doubleTapDashBoost*glm::min(accelerationSpeed, addSpeed), 50.0f);
                 velocity += accelerationSpeed * wishDirection;
                 ImGui::Text("velocity: (%f, %f, %f) |%f|", velocity.x, velocity.y, velocity.z, glm::length(velocity));
             }
 
-            if (controller->Jumping() && !controller->Crouching() && velocity.y == 0.f) {
+            //you cant jump and dash at the same time - since there is no friction in the air and we would thus dash much further in the air
+            if (!m_PlayerIsDashing && controller->Jumping() && !controller->Crouching() && velocity.y == 0.f) {
```
Comment (sippeangelo) [inline src/Game/Systems/PlayerMovementSystem.cpp]: Use input controller

File: src/Game/Systems/PlayerMovementSystem.cpp [hunk 4]
```diff
@@ -158,3 +164,41 @@ bool PlayerMovementSystem::OnPlayerSpawned(Events::PlayerSpawned& e)
 
     return true;
 }
+void PlayerMovementSystem::assaultDashCheck(glm::vec3 controllerMovement, double dt, bool isJumping) {
```
Comment (sippeangelo) [inline src/Game/Systems/PlayerMovementSystem.cpp]: None of this will work for multiple players

**Comment by sippeangelo** [comment]
Double-tap to dash is an awesome feature. Just add a config option in Input.ini so the player can choose to disable it.

Dash also needs to happen when pressing shift, which should be a shared key for all the class abilities like dash, sprint and shield. Right now the input command for shift is "sprint".

**Comment by verysecrethero** [comment]
Now works great! Updated the pull request info text. Could possibly put cooldown/etc constant variables in the Dash Component if wanted

File: resources/Schema/Components/Dash.xsd [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:t="types">
+	<xs:import schemaLocation="../Types.xsd" namespace="types"/>
+
+	<xs:element name="Dash">
```
Comment (sippeangelo) [inline resources/Schema/Components/Dash.xsd]: Rename to DashAbility or something similar so we can group it with other abilities more easily.

**Pull request merged**